### DISCUSSION TITLE
Accessing Python variables from JS side in current PyScript

### DISCUSSION THREAD

**Discussion by zmoon** [pyscript/pyscript#1904] [Q&A] [answered]
Before, I was able to do this in a JS `<script>` with `pyscript.interpreter.globals.get()`, but with the new `type="module"` pyscript, this is not (automatically at least) available anymore.

Update: Is the method at https://jeff.glass/post/pyscript-js-functions/#commentary the current recommended? And what about if using MicroPython instead of Pyodide?

**Comment by JeffersGlass** [comment, marked as answer]
If you know in advance which Python objects you'll need on the JavaScript side, I recommend [exporting only those objects from Python](https://pyscript.recipes/2023.11.1/js-integration/export-js/).

If you need access to arbitrary Python objects in JS at runtime (i.e. objects that you don't know the names of beforehand), you can still [export a reference to the Python globals](https://pyscript.recipes/2023.11.1/js-integration/import-py/).

Those recipes are identical for Micropython and Pyodide.

**Reply by zmoon** [reply]
Thanks, I didn't know about [pyscript.recipes](https://pyscript.recipes/), that's useful

**Comment by WebReflection** [comment]
what @JeffersGlass said but I wonder ...

> Before, I was able to do this in a JS <script> with pyscript.interpreter.globals.get()

were you relying the synchronous and blocking bootstrap and the global field leak before?

FYI this won't work with Workers case because the interpreter is elsewhere, but if you need it on the main thread and you bootstrap on main too, Jeff's solution with *hooks* import works best.

In any other case you can dispatch events or pollute the global or use *sync* in workers to pass along any JS/Python value you like in the process, but definitively we won't leak interpreters on the main nor pyscript global neither, as we have multiple interpreters now and, once again, when workers are used these don't exist on main.

**Reply by zmoon** [reply]
Thanks for these details.

> were you relying the synchronous and blocking bootstrap and the global field leak before?

I'm not sure, [this](https://github.com/zmoon/thrower-segovia-diagrams/blob/4d884bf7b60d1d4ec2bdd004ae8b124ba9b39c8b/t.html#L117) is what I was doing.

**Reply by WebReflection** [reply]
interesting ... why didn't you just pass `_addScaleOptions(scale_choices)` there instead of needing that indirect access to interpreter globals? :thinking: this way survives workers too and it's, imho, **the** way to go for these use cases, curious to know why that wasn't considered as an option, if I might ask this question, thanks!

**Reply by zmoon** [reply]
Thanks, I'll try to remember that next time, not sure why I didn't do that 😅. For [update](https://github.com/zmoon/thrower-segovia-diagrams/pull/5/files) I did it in PyScript instead, does that have the same benefits you mention?

**Reply by WebReflection** [reply]
Absolutely! Anything worth doing in Python *should* be done in Python and you remove a lot of roundtrips around the code.
